livre/lib/cache/HistogramObject.cpp

```diff
@@ -0,0 +1,240 @@
+                  const uint64_t scaleFactor) const
+    {
+        uint64_t* bins = histogram.getBins();
+        for( size_t i = padding.x(); i < blockSize.x() - padding.x(); ++i )
+            for( size_t j = padding.y(); j < blockSize.y() - padding.y(); ++j )
+                for( size_t k = padding.z(); k < blockSize.z() -padding.z(); ++k )
+                    for( size_t c = 0; c < compCount; ++c )
+                    {
+                        const size_t index = compCount * i * blockSize.y() * blockSize.z() +
+                                             compCount * j * blockSize.z() +
+                                             compCount * k + c;
+                        bins[ 128 + rawData[ index ]] += scaleFactor;
+                    }
+    }
+
+    void binData( const uint8_t* rawData,
+                  Histogram& histogram,
+                  const Vector3ui& blockSize,
+                  const Vector3ui& padding,
+                  const size_t compCount,
+                  const uint64_t scaleFactor ) const
+    {
+
+        uint64_t* bins = histogram.getBins();
+        for( size_t i = padding.x(); i < ( blockSize.x() - padding.x( )); ++i )
+            for( size_t j = padding.y(); j < ( blockSize.y() - padding.y( )); ++j )
+                for( size_t k = padding.z(); ( k < blockSize.z() -padding.z( )); ++k )
+                    for( size_t c = 0; c < compCount; ++c )
+                    {
+                        const size_t index = compCount * i * blockSize.y() * blockSize.z() +
+                                             compCount * j * blockSize.z() +
+                                             compCount * k + c;
+                        const uint8_t data = rawData[ index ];
+                        bins[ data ] += scaleFactor;
+                    }
+    }
+
+    template< class SRC_TYPE >
+    void binData( const SRC_TYPE* rawData,
+                  Histogram& histogram,
+                  const Vector3ui& blockSize,
+                  const Vector3ui& padding,
+                  const size_t compCount,
+                  const uint64_t scaleFactor ) const
+    {
+        uint64_t* bins = histogram.getBins();
+        const float minVal = std::numeric_limits< SRC_TYPE >::min();
+        const float maxVal = std::numeric_limits< SRC_TYPE >::max();
+        const float range = maxVal - minVal;
+
+        for( size_t i = padding.x(); i < blockSize.x() - padding.x(); ++i )
+            for( size_t j = padding.y(); j < blockSize.y() - padding.y(); ++j )
+                for( size_t k = padding.z(); k < blockSize.z() -padding.z(); ++k )
+                    for( size_t c = 0; c < compCount; ++c )
+                    {
+                        const size_t index = compCount * i * blockSize.y() * blockSize.z() +
+                                             compCount * j * blockSize.z() +
+                                             compCount * k + c;
+                        const float data = rawData[ index ];
```

Comment:
To avoid repeating the loop code twice you can keep a single template version of this function and implement the innermost part of the loop as three overloads, one as a template function which does range scale and the other two for the 8bit types.